### PR TITLE
Add non-moving allocator

### PR DESCRIPTION
**Pull request by anuraaga**

### CODE REVIEW

File: allocator/nonmoving_windows.go [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+	// Reserve max bytes of address space, to ensure we won't need to move it.
+	// This does not commit memory.
+	r, _, err := procVirtualAlloc.Call(0, uintptr(max), windows_MEM_RESERVE, windows_PAGE_READWRITE)
+	if err != nil {
+		panic(fmt.Errorf("allocator_windows: failed to reserve memory: %w", err))
+	}
+
+	buf := unsafe.Slice((*byte)(unsafe.Pointer(r)), int(max))
+	return &virtualMemory{buf: buf[:cap], addr: r}
```
Comment (ncruces) [inline allocator/nonmoving_windows.go]: I think this needs to be (compare with unix above):
```go
return &virtualMemory{buf: buf[:0], addr: r}
```

We're not commiting the `cap` bytes yet. Reallocate will do that, and this needs to know:
```go
com := uint64(len(m.buf))
```

File: allocator/nonmoving_unix.go [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+		// This ensures int(max) overflows to a negative value,
+		// and syscall.Mmap returns EINVAL.
+		max = math.MaxUint64
+	}
+
+	// Reserve max bytes of address space, to ensure we won't need to move it.
+	// A protected, private, anonymous mapping should not commit memory.
+	b, err := syscall.Mmap(-1, 0, int(max), syscall.PROT_NONE, syscall.MAP_PRIVATE|syscall.MAP_ANON)
+	if err != nil {
+		panic(err)
```
Comment (ncruces) [inline allocator/nonmoving_unix.go]: If we're wrapping windows errors, should we wrap unix errors too?

File: allocator/nonmoving_unix.go [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+	res := uint64(cap(m.buf))
+	if com < size && size < res {
+		// Round up to the page size.
+		rnd := uint64(syscall.Getpagesize() - 1)
+		new := (size + rnd) &^ rnd
+
+		// Commit additional memory up to new bytes.
+		err := syscall.Mprotect(m.buf[com:new], syscall.PROT_READ|syscall.PROT_WRITE)
+		if err != nil {
+			panic(err)
```
Comment (ncruces) [inline allocator/nonmoving_unix.go]: Ditto.

File: allocator/nonmoving_unix.go [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+	}
+	// Limit returned capacity because bytes beyond
+	// len(m.buf) have not yet been committed.
+	return m.buf[:size:len(m.buf)]
+}
+
+func (m *mmappedMemory) Free() {
+	err := syscall.Munmap(m.buf[:cap(m.buf)])
+	if err != nil {
+		panic(err)
```
Comment (ncruces) [inline allocator/nonmoving_unix.go]: Same.

**Review by ncruces — APPROVED** [review]

**Comment by anuraaga** [comment]
@ncruces Thanks for the review! Will add some tests separately, want to confirm some points

**Pull request merged**